File: sdks/swift/Sources/SdkTpl/SdkTplClient.swift
import Foundation

/// Response returned by the health check endpoint.
public struct HealthResponse: Codable, Sendable {
    public let status: String
    public let timestamp: String
    public let version: String?
}

/// Main client for the {{SDK_NAME}} Swift SDK.
///
/// Create an instance with a ``SdkTplConfig`` and use it to interact with
/// the {{SDK_NAME}} API.
///
/// ```swift
/// let client = SdkTplClient(config: SdkTplConfig(apiKey: "your-api-key"))
/// let health = try await client.healthCheck()
/// print(health.status)
/// client.close()
/// ```
public final class SdkTplClient: @unchecked Sendable {

    private let httpClient: HttpClient
    private let config: SdkTplConfig

    // MARK: - Initialisation

    /// Creates a new {{SDK_NAME}} API client.
    ///
    /// - Parameter config: The client configuration including the API key.
    /// - Throws: A ``SdkTplError`` with code ``ErrorCode/authMissingKey``
    ///   when the API key is empty.
    public init(config: SdkTplConfig) throws {
        guard !config.apiKey.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw SdkTplError(
                code: .authMissingKey,
                message: "API key is required"
            )
        }

        self.config = config
        self.httpClient = HttpClient(apiKey: config.apiKey, config: config)
    }

    // MARK: - Public API

    /// Performs a health check against the {{SDK_NAME}} API.
    ///
    /// Use this to verify connectivity and that the API key is valid before
    /// issuing business requests.
    ///
    /// - Returns: A ``HealthResponse`` describing the API status.
    /// - Throws: A ``SdkTplError`` on network or authentication failures.
    public func healthCheck() async throws -> HealthResponse {
        let data = try await httpClient.request(method: "GET", path: "/health")
        let decoder = JSONDecoder()
        return try decoder.decode(HealthResponse.self, from: data)
    }

    /// Returns a read-only snapshot of the current configuration with
    /// sensitive fields (API keys) omitted.
    public func getConfig() -> [String: Any] {
        return [
            "baseUrl": config.baseUrl,
            "timeout": config.timeout,
            "maxRetries": config.retryConfig.maxRetries,
            "failureThreshold": config.circuitBreakerConfig.failureThreshold,
            "enableRequestSigning": config.enableRequestSigning,
            "enableErrorSanitization": config.enableErrorSanitization,
        ]
    }

    /// Releases any resources held by the client.
    ///
    /// Call this when the client is no longer needed.
    public func close() {
        httpClient.close()
    }
}

File: sdks/php/src/SdkTplClient.php
<?php

declare(strict_types=1);

namespace SdkTpl;

use SdkTpl\Errors\SdkTplException;
use SdkTpl\Http\HttpClient;
use SdkTpl\Utils\Version;

/**
 * Main client for the {{SDK_NAME}} SDK.
 *
 * Usage:
 * ```php
 * $client = new SdkTplClient([
 *     'apiKey' => 'your-api-key',
 * ]);
 * $health = $client->healthCheck();
 * $client->close();
 * ```
 */
class SdkTplClient
{
    private SdkTplConfig $config;
    private HttpClient $httpClient;

    /**
     * @param array<string, mixed> $config Configuration options.
     *
     * @throws SdkTplException If the configuration is invalid.
     */
    public function __construct(array $config)
    {
        $this->config = new SdkTplConfig($config);
        $this->httpClient = new HttpClient($this->config);
    }

    /**
     * Performs a health check against the API.
     *
     * @return array{status: string, version: string, apiVersion: string}
     *
     * @throws SdkTplException If the request fails.
     */
    public function healthCheck(): array
    {
        try {
            $response = $this->httpClient->get('/health');

            return [
                'status' => 'ok',
                'version' => Version::SDK_VERSION,
                'apiVersion' => $response['apiVersion'] ?? 'unknown',
            ];
        } catch (SdkTplException $e) {
            throw $e;
        } catch (\Throwable $e) {
            throw SdkTplException::networkError('Health check failed: ' . $e->getMessage(), $e);
        }
    }

    /**
     * Sends a request to the API.
     *
     * @param string                    $method HTTP method (GET, POST, PUT, DELETE).
     * @param string                    $path   API endpoint path.
     * @param array<string, mixed>|null $body   Optional request body.
     *
     * @return array<string, mixed> Parsed JSON response.
     *
     * @throws SdkTplException If the request fails.
     */
    public function request(string $method, string $path, ?array $body = null): array
    {
        return match (strtoupper($method)) {
            'GET' => $this->httpClient->get($path),
            'POST' => $this->httpClient->post($path, $body),
            'PUT' => $this->httpClient->put($path, $body),
            'DELETE' => $this->httpClient->delete($path),
            default => throw SdkTplException::validationError("Unsupported HTTP method: {$method}"),
        };
    }

    /**
     * Returns the current configuration.
     */
    public function getConfig(): SdkTplConfig
    {
        return $this->config;
    }

    /**
     * Closes the underlying HTTP client and releases resources.
     */
    public function close(): void
    {
        $this->httpClient->close();
    }
}
